### ISSUE TITLE
Proposal: raise useful error message when capistrano-* gem is required incorrectly

### ISSUE THREAD

**Issue by mattbrictson** [capistrano/capistrano#1451]
One point of confusion with capistrano is the loading process for plugins. Many capistrano-* gems plug into capistrano by enhancing the `load:defaults` task. If this enhancement is done too late in capistrano's boot process, `load:defaults` has already been invoked. In this case the gem malfunctions because its setup process has been skipped.

Essentially the gem has to be required in the `Capfile`:

```ruby
require "capistrano/setup"
# plugins must be loaded here, after setup
```

If the user instead puts the require statement in `deploy.rb`, the plugin breaks, often in mysterious ways. See https://github.com/capistrano/rails/issues/37.

Now clearly this is a problem of not reading the documentation. But perhaps capistrano can make this mistake less likely or show a helpful error message in this scenario?

Here's an idea: in `capistrano/setup.rb`, alter the `load:defaults` task after invoking it, such that further attempts to enhance it will fail. Here is a very basic way to do it:

```ruby
invoke 'load:defaults'
Rake.application["load:defaults"].freeze
```

Now if a plugin loads out of order (i.e. in `deploy.rb`), it will immediately fail with `can't modify frozen Rake::Task`.

We could take this idea one step further and use `extend` to override `Rake::Task#enhance` to provide a more useful error message.

What do you think? Is this a worthwhile effort to make capistrano more forgiving/easier to troubleshoot?

**Comment by leehambley** [comment]
I think that's a great suggestion Matt, but I'd be concerned that it's a half-measure, depending on where plugins actually hook in.  Another step in the right direction, that is for sure. Your idea of `Rake::Task#enhance` to maybe check if the object is `frozen?` and raise an error message along the lines of "looks like your plugin is late to the party bro" might be worthwhile too.

Could you take a shot at a patch ?

**Comment by mattbrictson** [comment]
@leehambley I'll take a shot at a patch and open a separate PR.

Seeing that this is a "half-measure" (I agree), do you have any thoughts on what a more comprehensive solution would be? Perhaps a more formal plugin API? Right now plugin authoring is a bit mysterious; even enhancing `load:defaults` is more of an undocumented trick than an official policy.

**Comment by mattbrictson** [comment]
Closed via #1452.
